src/Workspaces/Core/Portable/Shared/Utilities/StringBreaker.cs

```diff
@@ -182,131 +182,172 @@ internal static class StringBreaker
+            while (current < length && IsLower(identifier[current]))
+            {
+                current++;
+            }
+
+            return new TextSpan(wordStart, current - wordStart);
+        }
+
+        private static TextSpan ScanNumber(string identifier, int length, int wordStart)
+        {
+            var current = wordStart + 1;
+            while (current < length && char.IsDigit(identifier[current]))
+            {
+                current++;
+            }
+
+            return TextSpan.FromBounds(wordStart, current);
         }
 
-        private static bool TransitionFromUpperToLower(string identifier, bool word, int index, int wordStart)
+        private static int SkipPunctuation(string identifier, int length, int wordStart)
         {
-            if (word)
+            while (wordStart < length)
             {
-                // Cases this supports:
-                // 1) IDisposable -> I, Disposable
-                // 2) UIElement -> UI, Element
-                // 3) HTMLDocument -> HTML, Document
-                //
-                // etc.
-                if (index != wordStart &&
-                    index + 1 < identifier.Length)
+                var ch = identifier[wordStart];
+                if (ch != '_' && char.IsPunctuation(ch))
                 {
-                    var currentIsUpper = char.IsUpper(identifier[index]);
-                    var nextIsLower = char.IsLower(identifier[index + 1]);
-                    if (currentIsUpper && nextIsLower)
-                    {
-                        // We have a transition from an upper to a lower letter here.  But we only
-                        // want to break if all the letters that preceded are uppercase.  i.e. if we
-                        // have "Foo" we don't want to break that into "F, oo".  But if we have
-                        // "IFoo" or "UIFoo", then we want to break that into "I, Foo" and "UI,
-                        // Foo".  i.e. the last uppercase letter belongs to the lowercase letters
-                        // that follows.  Note: this will make the following not split properly:
-                        // "HELLOthere".  However, these sorts of names do not show up in .Net
-                        // programs.
-                        for (int i = wordStart; i < index; i++)
-                        {
-                            if (!char.IsUpper(identifier[i]))
-                            {
-                                return false;
-                            }
-                        }
-
-                        return true;
-                    }
+                    wordStart++;
+                    continue;
                 }
+
+                break;
             }
 
-            return false;
+            return wordStart;
         }
 
-        private static bool TransitionFromLowerToUpper(string identifier, bool word, int index)
+        private static bool IsLower(char c)
         {
```

Comment:
Hot path.  Have fast path for ascii to avoid extra IsLatin check in char.IsLower